infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -72,32 +88,92 @@ public KubernetesNamespaceFactory(
     }
   }
 
-  private boolean hasNoPlaceholders(String namespaceName) {
-    return NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().noneMatch(namespaceName::contains);
+  private boolean hasPlaceholders(String namespaceName) {
+    return namespaceName != null
+        && NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().anyMatch(namespaceName::contains);
+  }
+
+  /** True if namespace is potentially created for the workspaces, false otherwise. */
+  protected boolean isCreatingNamespaces(String workspaceId) throws InfrastructureException {
+    // ...namespace    | ...namespace exists | ...namespace.default | creating?
+    // no-placeholders |       no            |       null           | error
+    // no-placeholders |       no            |   no-placeholders    | no
+    // no-placeholders |       no            |    placeholders      | yes
+    // no-placeholders |      yes            |       null           | no
+    // no-placeholders |      yes            |   no-placeholders    | no
+    // no-placeholders |      yes            |    placeholders      | no
+    //  placeholders   |       no            |        null          | error
+    //  placeholders   |       no            |   no-placeholders    | no
+    //  placeholders   |       no            |    placeholders      | yes
+    //  placeholders   |      yes            |        null          | yes
+    //  placeholders   |      yes            |   no-placeholders    | yes
+    //  placeholders   |      yes            |    placeholders      | yes
+    boolean legacyWithPlaceholders = isNullOrEmpty(namespaceName) || hasPlaceholders(namespaceName);
+    boolean legacyExists =
+        checkNamespaceExists(
+            resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));
+
+    if (isNullOrEmpty(defaultNamespaceName) && !legacyExists) {
+      throw new InfrastructureException(
+          "Cannot determine whether a new namespace and service account should be"
+              + " created for workspace %s. There is no pre-existing workspace namespace to be found using the legacy"
+              + " `che.infra.kubernetes.namespace` property yet the `che.infra.kubernetes.namespace.default` property"
+              + " is undefined.");
+    }
+
+    boolean defaultHasPlaceholders = hasPlaceholders(defaultNamespaceName);
+
+    return (legacyWithPlaceholders && legacyExists) || (!legacyExists && defaultHasPlaceholders);
   }
 
   /**
-   * True if namespace is predefined for all workspaces. False if each workspace will be provided
-   * with a new namespace or provided for each user when using placeholders.
+   * @return true if the namespaces are fully managed by Che (e.g. created, used and deleted), false
+   *     otherwise
    */
-  public boolean isPredefined() {
-    return isPredefined;
+  public boolean isManagingNamespaces(String workspaceId) throws InfrastructureException {
+    boolean canBeManaged =
```

Comment:
I'm trying to understand what `canBeManaged` means...
maybe it makes sense to add an explaining comment like `//we manage namespace only when <workspaceid> is used in legacy or new default namespace name` <- just example, I did not understand this if yet )

Reply:
I reworked this method, so hopefully it is cleaner now.